File: linkedlist/ll.js
class Node{
  constructor(data,next =null){
   this.data = data
   this.next = next
  }
}

class linkedlist {
    constructor(){
        this.head =null
    }
}
// insert at begiening
linkedlist.prototype.insertAtBegining = function (data){
  const newnode = new Node(data)
  this.head = newnode
}

// insert at end
linkedlist.prototype.insertatlast = function(data){
const newnode = new Node(data)
if(!this.head){
  this.head = newnode
}

let last = this.head  // false
while(!this.head){
    last = last.next
}
last.next = newnode

}
// insert at Anywhere

linkedlist.prototype.insertAfter = function (data, prevNode){
    if(!prevNode){
        console.log("please provide previous node");
    }

    const newnode = new Node(data,prevNode.next)
       prevNode.next = newnode

}


// delete an element from beginning
linkedlist.prototype.DeleteElementfromBeg= function(){
  if(!this.head){
    console.log("Linkedlist is Empty");
  }
  this.next = this.head
}

// delete element From last 
linkedlist.prototype.delElemFromLast= function(){
  if(!this.head){
    console.log("Linkedlist is Empty");
  }
  if(!this.head.next ){
    this.head = null   // if there is only one value i L.L.
    return
  }
  let slast = this.head
    while(this.slast.next.next){
      slast = slast.next
    }

  slast.next=null
  }

// delete element at a given key
 linkedlist.prototype.delFromAny= function(key){
  if(!this.head){
    console.log("Linkedlist is Empty");
  }
  // value at head
   if(this.head.data === key){
    this.head = this.head.next
   }

   let current = this.head

   while(current.next !== null){
    if(current.next.data === key){
      current.next = current.next.next
      return
    }
    current.next=current
   }

   console.log(`No node Found with key : ${key}`);
}

   // searching in L.L.
   linkedlist.prototype.Serc=function(key){
   let current = this.head
    while(current){
      if(current.data===key){
        return true
      }
      current = current.next
    }
    return false
   }
  
// Traversal in L.L
linkedlist.prototype.Trav=function(){
  let current = this.head
   let listv = []
   while(current){
    listv.push(current.data); // add data to array
    current = current.next // move to next node
   }
   console.log(listv.join(" -> "));
}
linkedlist.prototype.reverse=function(){
let current = this.head
let prev = null
let next = null
while(current){
    next = current.next
    current.next = prev
    prev = current
    current = next
}
this.head = prev

}

linkedlist.prototype.che=function(){
  let last = this.head
    while(last.next && last.next.next){
        if( last.next.next==this.head.next){
            return true
        }
        last=last.next
    }
    return false
}


let ll = new linkedlist
ll.insertAtBegining(32)
ll.insertAfter(12,32)
ll.insertAfter(21,2)
ll.insertatlast(23)
ll.insertatlast(34)

console.log(ll.Serc(12));
ll.Trav()
ll.reverse()
ll.Trav()

File: linkedlist/ll.py
import pathlib
class Node:
    def __init__(self,data,next = None):
       self.data = data
       self.next = next
    
class LinkedList:
    def __init__(self,head=None):
        self.head = None

    def insertElementAtBegin(self,ele):
        node = Node(ele,self.head)
        self.head = node
        return
    
    def insertAtLast(self,ele):
        node =  Node(ele)
        if(self.head == None):
            self.head = node
            return
        
        last = self.head 
        while(last.next != None):
            last = last.next
        last.next = node
        return
        
    
    def insertAny(self,ele,prevnode):
        if(prevnode.next != None):
            self.head = self.head.next
            return
        
        else:
            node = Node(ele,prevnode)
            prevnode.next = node
            return
    
    def DeleteFromBegin(self):
        if(self.head != None):
            self.head = self.head.next
            return
        

    
    def DeletFromEnd(self):
        if(self.head == None):
            print("LinkedList is Empty")
            return
            
            
        if(self.head.next == None):
            self.head = None
            return
        
        slast = self.head
        while(slast.next.next == None):
            slast.next = None
            
        slast=slast.next
        return

    def DeleteFromAny(self,key):
        if(self.head == None):
            print("LinkedList is Empty")
        
        if(self.head.data == key):
            self.head=self.head.next
            return
        
        current = self.head
        while(current.next != None ):
            if(current.next.data == key):
                current.next = current.next.next
                return
            
            current = current.next
        print("No Node Found")

         
    def Search(self,key):
        if(self.head == None):
            print("LinkedList is Empty")
        current = self.head
        if(current.data == key):
            print("Found At :",current)
        
        while(current.next != None):
            if(current.next.data == key):
                return True
            else:
                current= current.next
        
        print("Element is not present in LinkedList : ",key)
    
    def Traverse(self):
        current = self.head
        tr="->"
        lst = []
        while(current):
            (lst.append(str(current.data)))
            current = current.next
        print(tr.join(lst))
       
         
ll = LinkedList()
ll.insertElementAtBegin(32)
ll.insertAtLast(43)
ll.Traverse()
print(ll.Search(43))
# ll.DeleteFromBegin()
ll.Traverse()
        


          
        
        

        



        







        

